q.ml

```diff
@@ -431,6 +348,144 @@ let  div_2exp x n =
   else make_real x.num (Z.shift_left x.den n)
 
 
+type supported_base =
+  | B2 | B8 | B10 | B16
+
+let int_of_base = function
+  | B2 -> 2
+  | B8 -> 8
+  | B10 -> 10
+  | B16 -> 16
+
+(* The current implementation supports plain decimals, decimal points,
+   scientific notation ('e' or 'E' for base 10 litteral and 'p' or 'P'
+   for base 16), and fraction of integers (eg. 1/2). In particular it
+   accepts any numeric litteral -without underscores ('_')- accepted
+   by OCaml's lexer.
+   Restrictions:
+   - does not handle '_' as their removal should probably be common to
+     Z.of_string and Q.of_string
+   - exponents in scientific notation should fit on an integer
+   - scientific notation only available in hexa and decimal (as in OCaml) *)
+let of_string =
+  (* return a boolean (true for negative) and the next offset to read *)
+  let parse_sign s i j =
+    if j < i + 1
+    then false, i
+    else
+      match s.[i] with
+      | '-' -> true , i + 1
+      | '+' -> false, i + 1
+      | _ -> false ,i
+  in
+  (* return the base and the next offset to read *)
+  let parse_base s i j =
+    if j < i + 2
+    then B10, i
+    else
+      match s.[i],s.[i+1] with
+      | '0',('x'|'X') -> B16, i + 2
+      | '0',('o'|'O') -> B8, i + 2
+      | '0',('b'|'B') -> B2, i + 2
+      | _ -> B10, i
+  in
+  (* [find_in_string s ~pos ~last pred] find the first index in the string between [pos]
+     (inclusive) and [last] (exclusive) that satisfy the predicate [pred] *)
+  let find_in_string s ~pos ~last p =
+    let rec loop s ~pos:i ~last p =
+      if i = last
+      then None
+      else if p s.[i]
+      then Some i
+      else loop s ~pos:(i + 1) ~last p
+    in
+    loop s ~pos ~last p
```

Comment:
I would rather move this function outside of `of_string`, as it is completely generic.

Also: I don't see the need for the inner `loop` function; can't you just make `find_in_string` recursive?

Reply:
I've moved find_in_string out